sharding/notary/notary.go

```diff
@@ -151,7 +165,279 @@ func joinNotaryPool(client mainchain.Client) error {
+// by calling the DeregisterNotary function in the SMC.
+func leaveNotaryPool(client mainchain.Client) error {
+
+	if inPool, err := isAccountInNotaryPool(client); !inPool || err != nil {
+		if err != nil {
+			return err
+		}
+		return errors.New("account has not been able to be deposited in notary pool")
+	}
+
+	txOps, err := client.CreateTXOpts(nil)
+	if err != nil {
+		return fmt.Errorf("unable to create txOpts: %v", err)
+	}
+	tx, err := client.SMCTransactor().DeregisterNotary(txOps)
+
+	if err != nil {
+		return fmt.Errorf("unable to deregister notary: %v", err)
+	}
+
+	receipt, err := client.TransactionReceipt(tx.Hash())
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == types.ReceiptStatusFailed {
+		return errors.New("transaction was not successful, unable to deregister notary")
+	}
+
+	if dreg, err := hasAccountBeenDeregistered(client); !dreg || err != nil {
+		if err != nil {
+			return err
+		}
+		return errors.New("notary unable to be deregistered successfully from pool")
+	}
+
+	log.Info(fmt.Sprintf("Notary deregistered from the pool with hash: %s", tx.Hash().Hex()))
+	return nil
+
+}
+
+// releaseNotary releases the Notary from the pool by deleting the notary from
+// the registry and transferring back the deposit
+func releaseNotary(client mainchain.Client) error {
+	nreg, err := getNotaryRegistry(client)
+	if err != nil {
+		return err
+	}
+	if !nreg.Deposited {
+		return errors.New("account has not deposited in the Notary Pool")
+	}
+
+	if nreg.DeregisteredPeriod.Cmp(big.NewInt(0)) == 0 {
+
+		return errors.New("notary is still registered to the pool")
+
+	}
+
+	if lockup, err := isLockUpOver(client); !lockup || err != nil {
+		if err != nil {
+			return err
+		}
+		return errors.New("lockup period is not over")
+	}
+
+	txOps, err := client.CreateTXOpts(nil)
+	if err != nil {
+		return fmt.Errorf("unable to create txOpts: %v", err)
+	}
+	tx, err := client.SMCTransactor().ReleaseNotary(txOps)
+
+	if err != nil {
+		return fmt.Errorf("unable to Release Notary: %v", err)
+	}
+
+	receipt, err := client.TransactionReceipt(tx.Hash())
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == types.ReceiptStatusFailed {
+
+		return errors.New("transaction was not successful, unable to release Notary")
+
+	}
+
+	nreg, err = getNotaryRegistry(client)
+
+	if err != nil {
+
+		return err
+
+	}
+
+	if nreg.Deposited || nreg.Balance.Cmp(big.NewInt(0)) != 0 || nreg.DeregisteredPeriod.Cmp(big.NewInt(0)) != 0 || nreg.PoolIndex.Cmp(big.NewInt(0)) != 0 {
+
+		return errors.New("notary unable to be released from the pool")
+
+	}
+
+	log.Info(fmt.Sprintf("notary with address: %s released from pool", client.Account().Address.Hex()))
+
+	return nil
+
+}
+
+// submitVote votes for a collation on the shard
+// by taking in the shard and the hash of the collation header
+func submitVote(shard sharding.Shard, client mainchain.Client, headerHash *common.Hash) error {
```

Comment:
This method is over 100 lines and seems to be doing many things. 
Can you refactor this into smaller responsibility methods?

Submit vote should submit a vote, block until the vote tx has been mined, return error if the tx failed.

This methods seems to be doing much more than that:
- Checking the last submitted collation
- Checking the notary is deposited
- Checking the notary index
- Checking collation record?
- Checking chunk root
- Checking the notary has not already voted
- Checking the notary is in committee
- Submitting the vote
- Immediately getting the transaction receipt (not waiting for the transaction to be mined?)
- Checking again if the notary has voted 
- Logging that the notary has voted
- Checking collation records again
- Setting the canonical chain!!

Please read about the SOLID principals https://en.wikipedia.org/wiki/SOLID